File: server/src/services/documents/documentService.ts
const pdfParse = require('pdf-parse');

export interface DocumentChunk {
  chunkIndex: number;
  text: string;
  charStart: number;
  charEnd: number;
}

export interface DocumentMeta {
  id: string;
  type: 'resume' | 'job_description';
  filename: string;
  fileSize: number;
  uploadedAt: string;
  pageCount: number;
  extractedText: string;
  charCount: number;
  chunks: DocumentChunk[];
}

class DocumentService {
  private activeResume: DocumentMeta | null = null;
  private activeJobDescription: DocumentMeta | null = null;

  // ---------- RAG Chunking ----------
  private chunkText(text: string, chunkSize = 600, overlap = 100): DocumentChunk[] {
    const chunks: DocumentChunk[] = [];
    let start = 0;
    let index = 0;

    while (start < text.length) {
      const end = Math.min(start + chunkSize, text.length);
      chunks.push({
        chunkIndex: index++,
        text: text.slice(start, end),
        charStart: start,
        charEnd: end,
      });
      if (end === text.length) break;
      start += chunkSize - overlap;
    }
    return chunks;
  }

  // ---------- TF-IDF Style Keyword Retrieval ----------
  retrieveRelevantChunks(
    doc: DocumentMeta,
    query: string,
    topK = 3
  ): DocumentChunk[] {
    if (!doc.chunks.length) return [];

    const queryTokens = query.toLowerCase().split(/\W+/).filter(Boolean);

    const scored = doc.chunks.map((chunk) => {
      const chunkLower = chunk.text.toLowerCase();
      const score = queryTokens.reduce((acc, token) => {
        const matches = (chunkLower.match(new RegExp(token, 'g')) || []).length;
        return acc + matches;
      }, 0);
      return { chunk, score };
    });

    return scored
      .sort((a, b) => b.score - a.score)
      .slice(0, topK)
      .map((s) => s.chunk);
  }

  // ---------- Build RAG Context String ----------
  buildRagContext(query: string): string {
    const resume = this.activeResume;
    const jd = this.activeJobDescription;
    let context = '';

    if (resume) {
      const chunks = this.retrieveRelevantChunks(resume, query);
      context += `\n--- RESUME CONTEXT (${resume.filename}) ---\n`;
      context += chunks.map((c) => c.text).join('\n...\n');
    }

    if (jd) {
      const chunks = this.retrieveRelevantChunks(jd, query);
      context += `\n--- JOB DESCRIPTION CONTEXT (${jd.filename}) ---\n`;
      context += chunks.map((c) => c.text).join('\n...\n');
    }

    return context;
  }

  // ---------- PDF Parsing ----------
  async parsePdfBuffer(buffer: Buffer, filename: string, type: 'resume' | 'job_description'): Promise<DocumentMeta> {
    try {
      const parsed = await pdfParse(buffer);
      const text = (parsed.text || '').trim();
      const chunks = this.chunkText(text);

      const doc: DocumentMeta = {
        id: `${type}_${Date.now()}`,
        type,
        filename,
        fileSize: buffer.length,
        uploadedAt: new Date().toISOString(),
        pageCount: parsed.numpages || 1,
        extractedText: text,
        charCount: text.length,
        chunks,
      };

      if (type === 'resume') {
        this.activeResume = doc;
      } else {
        this.activeJobDescription = doc;
      }

      return doc;
    } catch (err: any) {
      throw new Error(`Failed to parse PDF file "${filename}": ${err.message}`);
    }
  }

  setPlainTextJd(text: string, title = 'Job_Description.txt'): DocumentMeta {
    const trimmed = text.trim();
    const chunks = this.chunkText(trimmed);
    const doc: DocumentMeta = {
      id: `job_description_${Date.now()}`,
      type: 'job_description',
      filename: title,
      fileSize: Buffer.byteLength(trimmed, 'utf8'),
      uploadedAt: new Date().toISOString(),
      pageCount: 1,
      extractedText: trimmed,
      charCount: trimmed.length,
      chunks,
    };
    this.activeJobDescription = doc;
    return doc;
  }

  getResume(): DocumentMeta | null {
    return this.activeResume;
  }

  getJobDescription(): DocumentMeta | null {
    return this.activeJobDescription;
  }

  deleteDocument(type: 'resume' | 'job_description'): boolean {
    if (type === 'resume') {
      this.activeResume = null;
      return true;
    } else if (type === 'job_description') {
      this.activeJobDescription = null;
      return true;
    }
    return false;
  }

  getStatus() {
    return {
      resume: this.activeResume
        ? {
            id: this.activeResume.id,
            filename: this.activeResume.filename,
            fileSize: this.activeResume.fileSize,
            uploadedAt: this.activeResume.uploadedAt,
            pageCount: this.activeResume.pageCount,
            charCount: this.activeResume.charCount,
            chunkCount: this.activeResume.chunks.length,
            preview: this.activeResume.extractedText.slice(0, 400) + '...',
          }
        : null,
      jobDescription: this.activeJobDescription
        ? {
            id: this.activeJobDescription.id,
            filename: this.activeJobDescription.filename,
            fileSize: this.activeJobDescription.fileSize,
            uploadedAt: this.activeJobDescription.uploadedAt,
            pageCount: this.activeJobDescription.pageCount,
            charCount: this.activeJobDescription.charCount,
            chunkCount: this.activeJobDescription.chunks.length,
            preview: this.activeJobDescription.extractedText.slice(0, 400) + '...',
          }
        : null,
    };
  }
}

export const documentService = new DocumentService();

File: python_server/app/services/document_service.py
import re
from typing import Any, Dict, List, Optional, Tuple
from uuid import uuid4

from pypdf import PdfReader


class DocumentChunk:
    def __init__(self, chunk_index: int, text: str, char_start: int, char_end: int):
        self.chunk_index = chunk_index
        self.text = text
        self.char_start = char_start
        self.char_end = char_end


class DocumentMeta:
    def __init__(
        self,
        doc_type: str,
        filename: str,
        file_size: int,
        extracted_text: str,
        page_count: int,
        uploaded_at: str,
    ):
        self.id = f"{doc_type}_{uuid4().hex[:8]}"
        self.type = doc_type
        self.filename = filename
        self.file_size = file_size
        self.extracted_text = extracted_text
        self.char_count = len(extracted_text)
        self.page_count = page_count
        self.uploaded_at = uploaded_at
        self.chunks = self._chunk_text(extracted_text)

    def _chunk_text(self, text: str, chunk_size: int = 600, overlap: int = 100) -> List[DocumentChunk]:
        chunks: List[DocumentChunk] = []
        start = 0
        index = 0
        while start < len(text):
            end = min(start + chunk_size, len(text))
            chunk_text = text[start:end]
            chunks.append(DocumentChunk(index, chunk_text, start, end))
            index += 1
            if end == len(text):
                break
            start += chunk_size - overlap
        return chunks


class DocumentService:
    def __init__(self):
        self.active_resume: Optional[DocumentMeta] = None
        self.active_job_description: Optional[DocumentMeta] = None

    def retrieve_relevant_chunks(self, doc: DocumentMeta, query: str, top_k: int = 3) -> List[DocumentChunk]:
        if not doc.chunks:
            return []

        query_tokens = [token.lower() for token in re.split(r"\W+", query) if token.strip()]
        scored: List[Tuple[DocumentChunk, int]] = []
        for chunk in doc.chunks:
            chunk_lower = chunk.text.lower()
            score = 0
            for token in query_tokens:
                score += len(re.findall(re.escape(token), chunk_lower))
            scored.append((chunk, score))

        scored.sort(key=lambda item: item[1], reverse=True)
        return [chunk for chunk, _ in scored[:top_k]]

    def build_rag_context(self, query: str) -> str:
        context_parts: List[str] = []

        if self.active_resume:
            chunks = self.retrieve_relevant_chunks(self.active_resume, query)
            context_parts.append(f"\n--- RESUME CONTEXT ({self.active_resume.filename}) ---\n")
            context_parts.append("\n...\n".join(chunk.text for chunk in chunks))

        if self.active_job_description:
            chunks = self.retrieve_relevant_chunks(self.active_job_description, query)
            context_parts.append(f"\n--- JOB DESCRIPTION CONTEXT ({self.active_job_description.filename}) ---\n")
            context_parts.append("\n...\n".join(chunk.text for chunk in chunks))

        return "".join(context_parts)

    def parse_pdf_buffer(self, buffer: bytes, filename: str, doc_type: str) -> DocumentMeta:
        try:
            reader = PdfReader(buffer)
            text_parts: List[str] = []
            for page in reader.pages:
                text = page.extract_text() or ""
                text_parts.append(text)
            extracted = "\n".join(text_parts).strip()
            doc = DocumentMeta(doc_type, filename, len(buffer), extracted, len(reader.pages), __import__("datetime").datetime.utcnow().isoformat())
            if doc_type == "resume":
                self.active_resume = doc
            else:
                self.active_job_description = doc
            return doc
        except Exception as exc:  # pragma: no cover
            raise ValueError(f'Failed to parse PDF file "{filename}": {str(exc)}') from exc

    def set_plain_text_jd(self, text: str, title: str = "Job_Description.txt") -> DocumentMeta:
        cleaned = text.strip()
        doc = DocumentMeta("job_description", title, len(cleaned.encode("utf-8")), cleaned, 1, __import__("datetime").datetime.utcnow().isoformat())
        self.active_job_description = doc
        return doc

    def get_resume(self) -> Optional[DocumentMeta]:
        return self.active_resume

    def get_job_description(self) -> Optional[DocumentMeta]:
        return self.active_job_description

    def delete_document(self, doc_type: str) -> bool:
        if doc_type == "resume":
            self.active_resume = None
            return True
        if doc_type == "job_description":
            self.active_job_description = None
            return True
        return False

    def get_status(self) -> Dict[str, Any]:
        return {
            "resume": self._serialize_doc(self.active_resume),
            "jobDescription": self._serialize_doc(self.active_job_description),
        }

    def _serialize_doc(self, doc: Optional[DocumentMeta]) -> Optional[Dict[str, Any]]:
        if not doc:
            return None
        return {
            "id": doc.id,
            "filename": doc.filename,
            "fileSize": doc.file_size,
            "uploadedAt": doc.uploaded_at,
            "pageCount": doc.page_count,
            "charCount": doc.char_count,
            "chunkCount": len(doc.chunks),
            "preview": (doc.extracted_text[:400] + "...") if len(doc.extracted_text) > 400 else doc.extracted_text,
        }


document_service = DocumentService()
